### DISCUSSION TITLE
Scrollbar thumb cannot be opaque?

### DISCUSSION THREAD

**Discussion by zambetti** [zed-industries/zed#9490] [Feature Requests]
It seems that scrollbar-area markers are drawn with a z-index or stacking order below the scrollbar thumb. Because of this, if the theme declares an opaque scrollbar thumb, the scrollbar-area markers are no longer visible when they are under the thumb (for example, markers indicating the scroll positions of search matches).

Prima facie, these markers could be drawn above/on-top-of/after the scrollbar thumb, to allow the scrollbar thumb to be opaque.

An opaque scrollbar thumb would also correct an distracting unintentional visual caused by the semi-transparent scrollbar thumb moving in one direction while the bright spot, where the current line highlight intersects the scrollbar thumb, moves in the opposite direction underneath it (screenshot below, intersection pointed to by pink arrow)

Does anyone know if it might be possible to reorder the stacking to allow the scrollbar thumb to be drawn before the markers?

Related to this, it seems the colors for the markers in the scrollbar area do not match the color specified in the search.match_background, does anyone know what the name/key is for styling the scrollbar markers in a theme?

![intersection](https://github.com/zed-industries/zed/assets/98515/b95a4b76-cdee-4e07-aefd-9b887433e662)

**Comment by alygin** [comment]
> Does anyone know if it might be possible to reorder the stacking to allow the scrollbar thumb to be drawn before the markers?

It's not possible via theming, though it's pretty easy to fix in the code.

> what the name/key is for styling the scrollbar markers in a theme?

It's the `info` color for search results and matching symbols.

**Reply by zambetti** [reply]
> It's the `info` color for search results and matching symbols.

Thank you for this. This is exactly what I was searching for.

**Reply by zambetti** [reply]
> It's not possible via theming, though it's pretty easy to fix in the code.

I'm not certain, but `paint_scrollbar()` in zed/crates/editor/src/element.rs might be the place to change the draw order. The last `cx.paint_quad()` looks to be the scrollbar thumb itself. And (all?) markers are drawn before. Most of those markers seem to use the `cx.theme().status().info` color as you mentioned. I would guess that moving that last `cx.paint_quad()` which paints the thumb, to happen earlier in the function should re-order the stacking.

**Reply by alygin** [reply]
You're right, that'll do the job. Though I'm not sure it's a good thing to do :)  In certain cases it may lead to the effect when the thumb is fully burried under markers and invisible to the user.

Is there something that prevents you from making the thumb semi-transparent?

**Reply by zambetti** [reply]
> In certain cases it may lead to the effect when the thumb is fully burried under markers and invisible to the user.

This could be handled by specifying semi-transparent colors for the markers, rather than forcing the scrollbar thumb to be semi-transparent.
 
> Is there something that prevents you from making the thumb semi-transparent?

Having a transparent thumb causes messy visuals like this (screenshot below). Here the current/highlighted line is showing through the semi-transparent thumb, and the marker visibility is reduced by the text underneath.

![Screenshot 2024-03-19 at 14 05 04](https://github.com/zed-industries/zed/assets/98515/5ef3eb73-3c56-4b64-a374-db6ecdaab716)

I suppose an opaque scrollbar track might be the easiest solution for now.

Thinking about this more, perhaps the design of Zed's scroll thumb could also be improved. If it were skinnier than the track, markers would extend past it horizontally, and thus always be at least somewhat visible no matter the theme. [intellij](https://jetbrains.design/intellij/controls/scrollbar/) takes this approach which is quite clean.

![scrollbar-detailed@2x](https://github.com/zed-industries/zed/assets/98515/ec0e4e4c-d282-4df5-82a9-ebc6e78403e0)

Another thing to perhaps consider here are blend-modes to ensure marker visibility at all times. Opacity alone is likely not enough considering that any colors could be underneath the scrollbar area.

**Reply by alygin** [reply]
> This could be handled by specifying semi-transparent colors for the markers, no?

Yes, but then they'll always be bleak, not just in corner cases.

> Having a transparent thumb causes messy visuals like this (screenshot below).

Interesting, I never payed attention to that. But it looks like all other editors do the same.

Actually, I don't know the best way to organize scrollbars and markers, but I have also been trying to improve how it looks and behaves. You might be interested in joining discussions here:
- #9070 
- #9080

**Reply by zambetti** [reply]
> You might be interested in joining discussions here:

I was not aware that the scrollbar/area was under such active development, great to hear of your work on this. I will follow the discussion there. Thank you.

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/support
